### PR TITLE
Bookmarks API

### PR DESCRIPTION
**Pull request by oyiptong**
This contains the bookmarks API for requests and updates

### CODE REVIEW

**Comment by oyiptong** [comment]
r? @marcoscaceres or @mzhilyaev 

Also, this may fix the concurrency issue in #93

**Comment by oyiptong** [comment]
@k88hudson please check the data types sent to content. It's a bit different than what you're expecting and that's due to how bookmarks are implemented.

File: lib/PlacesProvider.js [hunk 1]
```diff
@@ -158,28 +190,117 @@ Links.prototype = {
...
-                   " LEFT JOIN moz_favicons " +
-                   " ON favicon_id = moz_favicons.id " +
-                   " WHERE hidden = 0 AND last_visit_date NOTNULL " +
-                   " ORDER BY rev_host, frecency, last_visit_date, moz_places.url DESC " +
-                   ")    " +
-                   "GROUP BY rev_host " +
-                   "ORDER BY frecency DESC, lastVisitDate DESC, url " +
-                   "LIMIT :limit";
+    let sqlQuery = `SELECT url, title, frecency,
+                          last_visit_date as lastVisitDate, favicon, mime_type,
```
Comment (pdehaan) [inline lib/PlacesProvider.js]: Nit: this seems a bit confusing that we're mixing some `lastVisitDate` camelCase with `mime_type` underscores. Not sure if it's worth standardizing on one vs the other [in a future PR].

File: lib/PlacesProvider.js [hunk 2]
```diff
@@ -158,28 +190,117 @@ Links.prototype = {
...
+    }
+
+    let sqlQuery = `SELECT p.url, p.title, p.frecency,
+                          p.last_visit_date as lastVisitDate,
+                          b.lastModified,
+                          b.id as bookmarkId,
+                          b.title as bookmarkTitle,
+                          b.guid as bookmarkGuid,
+                          "bookmark" as type,
+                          f.data as favicon, f.mime_type
```
Comment (pdehaan) [inline lib/PlacesProvider.js]: Same as above. Everything else is mixedCase apart from poor ol' `mime_type`.

File: lib/PlacesProvider.js [hunk 3]
```diff
@@ -115,7 +118,38 @@ Links.prototype = {
...
+    },
+
+    onItemRemoved(id, folderId, index, type, uri) {
+      if (type == Bookmarks.TYPE_BOOKMARK) {
+        gLinks.emit("bookmarkRemoved", {bookmarkId: id, url: uri.spec});
+      }
+    },
+
+    onItemChanged(id, property, isAnnotation, value, lastModified, type) {
+      if (type == Bookmarks.TYPE_BOOKMARK) {
```
Comment (pdehaan) [inline lib/PlacesProvider.js]: Not sure if we want to change lines 130, 138, and 144 above to use `===` vs it's sketchy cousin `==`.

**Note to self:** check why our current ESLint rules didn't catch this, and possibly look into adopting Airbnb style guide or something. :suspect:

Comment (oyiptong) [inline lib/PlacesProvider.js]: Ah! We need to make sure we catch that. We're using the Mozilla style guide, but it's very lax, to make sure gecko at least has a chance of passing linting

Comment (pdehaan) [inline lib/PlacesProvider.js]: Filed #112; re: bulletproofing our .eslintrc file.

**Comment by k88hudson** [comment]
LGTM after those nits

File: lib/PlacesProvider.js (hunk 1 above)
Comment (oyiptong) [inline lib/PlacesProvider.js]: Good idea. I'll make the change in this PR

Comment (pdehaan) [inline lib/PlacesProvider.js]: ESLint has a [`camelcase`](http://eslint.org/docs/rules/camelcase.html) rule which causes 20 new errors in our code (yet none of them are related to this `mime_type` business).

Grep solves the day with this list, but I'm not sure what else out there may have odd underscores in the name that ESLint isn't catching:
```
✗ grep -irn "mime_type" . --exclude-dir=node_modules
./lib/PlacesProvider.js:162:                   "       last_visit_date as lastVisitDate, favicon, mime_type, " +
./lib/PlacesProvider.js:166:                   " SELECT rev_host, moz_places.url, moz_favicons.data as favicon, mime_type, title, frecency, last_visit_date " +
./lib/PlacesProvider.js:178:                  columns: ["url", "title", "lastVisitDate", "frecency", "favicon", "mime_type", "type"],
./lib/PlacesProvider.js:185:        link.favicon = `data:${link.mime_type};base64,${encodedData}`;
./lib/PlacesProvider.js:187:      delete link.mime_type;
```

**Comment by oyiptong** [comment]
@k88hudson I added the filters for folders

**Pull request merged**